### PR TITLE
WRN-18729: Added `editable` prop to Scroller to support editing items

### PR DESCRIPTION
**Pull request by MikyungKim**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Added `editable` prop to Scroller to support editing items

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
To support editing, added a flex wrapper and used flex `order` to order items. 

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
WRN-18729

### Comments
Enact-DCO-1.0-Signed-off-by: Mikyung Kim (mikyung27.kim@lge.com)

### CODE REVIEW

**Comment by jeonghee27** [comment]
I got error when I scroll with hoverToScroll.
```
encountered two children with the same key, `1`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
    at div
    at EditableWrapper (http://10.178.92.42:8080/main.iframe.bundle.js:53256:5)
    at div
    at ScrollerBasic (http://10.178.92.42:8080/main.iframe.bundle.js:20372:5)
    at div
    at http://10.178.92.42:8080/main.iframe.bundle.js:33337:5
    at http://10.178.92.42:8080/main.iframe.bundle.js:24497:9
    at Scroller (http://10.178.92.42:8080/main.iframe.bundle.js:53705:19)
...
```

File: Scroller/EditableWrapper.module.less [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// EditableWrapper.module.less
+//
+
+.wrapper {
+	display: flex;
+	flex-direction: row;
+	min-width: fit-content;
+    padding-top: 60px;
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.module.less:8]:
```suggestion
padding-top: 60px;
```

File: Scroller/EditableWrapper.module.less [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.module.less:9]:
```suggestion
	min-width: fit-content;
	padding-top: 60px;
}
```

File: Scroller/EditableWrapper.module.less [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+.selected {
+	transition: transform 360ms;
+	transform: translateY(-48px);
+	will-change: transform;
+	z-index: 1;
+}
+
+.selectedTransform {
+	transform: translate(var(--selected-item-offset), -48px);
+}
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.module.less:24]: Let's unify with selected

```suggestion
.selected {
	transition: transform 360ms;
	transform: translate(var(--selected-item-offset), -48px);
	transform: translateY(-48px);
	will-change: transform;
	z-index: 1;
	transform: translate(var(--selected-item-offset), -48px);
}
```

File: Scroller/EditableWrapper.module.less [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+.rearranged {
+	transition: transform 360ms;
+}
+
+.rearrangedTransform {
+	transform: translateX(calc((-1) * var(--move-direction) * var(--item-width)));
+}
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.module.less:32]: Same. 

m
```suggestion
.rearranged {
	transition: transform 360ms;
	transform: translateX(calc((-1) * var(--move-direction) * var(--item-width)));
}
```

File: Scroller/EditableWrapper.js [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+		if (rearrangedItems.length > 0) {
+			let selectedOrder = selectedItem.style.order;
+			const changedOrder = [];
+
+			rearrangedItems.forEach((item) => {
+				const order = Number(item.style.order);
+				selectedOrder = order;
+				item.style.order = order - lastMoveDirection;
+				item.classList.remove(css.rearrangedTransform);
+				item.classList.remove(css.rearranged);
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.js:77]:
```suggestion
				item.classList.remove(css.rearranged);
```

File: Scroller/EditableWrapper.js [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+				// Set the selected item's offset to css variable
+				const offset = (toIndex - fromIndex) * itemWidth;
+				wrapperRef.current.style.setProperty('--selected-item-offset', offset + 'px');
+
+				// Reset addable flag to true
+				mutableRef.current.addable = true;
+
+				// If the current toIndex is new,
+				if (toIndex !== prevToIndex) {
+					// Determine the direcion of move from the latest from index
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.js:171]:
```suggestion
					// Determine the direction of move from the latest from index
```

File: Scroller/EditableWrapper.js [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+					const moveDirection = Math.sign(toIndex - prevToIndex);
+					// If the direction is changed and there are rearranged items, we remove them first.
+					if (lastMoveDirection && moveDirection !== lastMoveDirection && rearrangedItems.length > 0) {
+						mutableRef.current.addable = false;
+					}
+					if (mutableRef.current.addable) {
+						moveSiblings({moveDirection, toIndex});
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.js:178]: rename to addRearrangedItems

File: Scroller/EditableWrapper.js [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+					} else {
+						const numToRemove = moveDirection > 0 ? toIndex - prevToIndex : prevToIndex - toIndex;
+						if (rearrangedItems.length > 0) {
+							for (let i = 0; i < numToRemove; i++) {
+								const toItem = rearrangedItems.pop();
+								toItem?.classList.remove(css.rearrangedTransform);
+							}
+						}
```
Comment (jeonghee27) [inline Scroller/EditableWrapper.js:186]: Let's make to function. removeRearrangedItems

File: Scroller/Scroller.js [hunk 9]
```diff
@@ -88,7 +107,12 @@ let Scroller = ({'aria-label': ariaLabel, hoverToScroll, ...rest}) => {
 		<ResizeContext.Provider {...resizeContextProps}>
 			<ScrollContentWrapper {...scrollContainerProps} {...scrollContentWrapperRest}>
 				<ScrollBody {...focusableBodyProps}>
-					<UiScrollerBasic {...themeScrollContentProps} aria-label={ariaLabel} id={id} ref={scrollContentHandle} />
+					<UiScrollerBasic {...themeScrollContentProps} aria-label={ariaLabel} id={id} ref={scrollContentHandle}>
+						{(editable == null || direction !== 'horizontal') ?
+							children :
+							<EditableWrapper {...editableWrapperProps} />
```
Comment (jeonghee27) [inline Scroller/Scroller.js:113]:
```suggestion
						{(editable && direction === 'horizontal') ?
							<EditableWrapper {...editableWrapperProps} /> :
							children
```

File: Scroller/EditableWrapper.js [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+ * @class EditableWrapper
+ * @memberof sandstone/Scroller
+ * @ui
+ * @public
+ */
+const EditableWrapper = (props) => {
+	const {children, editable, scrollContainerHandle, scrollContentRef} = props;
+	let centered = true;
+	if (editable.centered != null) {
+		centered = editable.centered;
```
Comment (0x64) [inline Scroller/EditableWrapper.js:22]: Let's make it a single line.

File: Scroller/EditableWrapper.js [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+	}
```
Comment (MikyungKim) [inline Scroller/EditableWrapper.js:23]:
```suggestion
	let centerd = editable.centered != null ? editable.centered : true;
```

File: Scroller/EditableWrapper.js [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+			const orders = finalizeOrders();
+			forwardCustom('onComplete', () => ({orders}))({}, editable);
+			reset();
+		} else {
+			// Start editing by adding selected transition to selected item
+			startEditing(ev.target.parentElement);
+		}
+	}, [editable, finalizeOrders, reset, startEditing]);
+
+	// Move siblings
```
Comment (MikyungKim) [inline Scroller/EditableWrapper.js:123]:
```suggestion
	// Add rearranged items
```

File: Scroller/EditableWrapper.js [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+					if (lastMoveDirection && moveDirection !== lastMoveDirection && rearrangedItems.length > 0) {
+						const numToRemove = moveDirection > 0 ? toIndex - prevToIndex : prevToIndex - toIndex;
+						removeRearrangedItems(numToRemove);
+
+						// When there's jump, meaning, numToRemove is bigger than 0, we need to add an item
+						if (numToRemove > rearrangedItems.length) {
+							addRearrangedItems({moveDirection, toIndex});
+						}
+					} else {
+						addRearrangedItems({moveDirection, toIndex}); // addRearrangedItems
```
Comment (MikyungKim) [inline Scroller/EditableWrapper.js:185]:
```suggestion
						addRearrangedItems({moveDirection, toIndex});
```

**Review by jeonghee27 — APPROVED** [review]
LGTM

**Pull request merged**